modules/apps/29-fee/ibc_module.go

```diff
@@ -0,0 +1,263 @@
+	// that causes module account to have insufficient funds to refund
+	// all escrowed fees on the channel.
+	// Disable all channels to allow for coordinated fix to the issue
+	// and mitigate/reverse damage.
+	// NOTE: Underlying application's packets will still go through, but
+	// fee module will be disabled for all channels
+	if err != nil {
+		im.keeper.DisableAllChannels(ctx)
+	}
+	return im.app.OnChanCloseInit(ctx, portID, channelID)
+}
+
+// OnChanCloseConfirm implements the IBCModule interface
+func (im IBCModule) OnChanCloseConfirm(
+	ctx sdk.Context,
+	portID,
+	channelID string,
+) error {
+	// delete fee enabled on channel
+	// and refund any remaining fees escrowed on channel
+	im.keeper.DeleteFeeEnabled(ctx, portID, channelID)
+	err := im.keeper.RefundFeesOnChannel(ctx, portID, channelID)
+	// error should only be non-nil if there is a bug in the code
+	// that causes module account to have insufficient funds to refund
+	// all escrowed fees on the channel.
+	// Disable all channels to allow for coordinated fix to the issue
+	// and mitigate/reverse damage.
+	// NOTE: Underlying application's packets will still go through, but
+	// fee module will be disabled for all channels
+	if err != nil {
+		im.keeper.DisableAllChannels(ctx)
+	}
+	return im.app.OnChanCloseConfirm(ctx, portID, channelID)
+}
+
+// OnRecvPacket implements the IBCModule interface.
+// If fees are not enabled, this callback will default to the ibc-core packet callback
+func (im IBCModule) OnRecvPacket(
+	ctx sdk.Context,
+	packet channeltypes.Packet,
+	relayer sdk.AccAddress,
+) exported.Acknowledgement {
+	if !im.keeper.IsFeeEnabled(ctx, packet.DestinationPort, packet.DestinationChannel) {
+		return im.app.OnRecvPacket(ctx, packet, relayer)
+	}
+
+	ack := im.app.OnRecvPacket(ctx, packet, relayer)
+
+	forwardRelayer, found := im.keeper.GetCounterpartyAddress(ctx, relayer.String())
+
+	// incase of async aknowledgement (ack == nil) store the ForwardRelayer address for use later
+	if ack == nil && found {
+		im.keeper.SetForwardRelayerAddress(ctx, channeltypes.NewPacketId(packet.GetSourceChannel(), packet.GetSourcePort(), packet.GetSequence()), forwardRelayer)
+	}
+
+	return types.NewIncentivizedAcknowledgement(forwardRelayer, ack.Acknowledgement())
+}
+
+// OnAcknowledgementPacket implements the IBCModule interface
+// If fees are not enabled, this callback will default to the ibc-core packet callback
+func (im IBCModule) OnAcknowledgementPacket(
+	ctx sdk.Context,
+	packet channeltypes.Packet,
+	acknowledgement []byte,
+	relayer sdk.AccAddress,
+) error {
+	if !im.keeper.IsFeeEnabled(ctx, packet.SourcePort, packet.SourceChannel) {
+		return im.app.OnAcknowledgementPacket(ctx, packet, acknowledgement, relayer)
+	}
+
+	ack := new(types.IncentivizedAcknowledgement)
+	if err := types.ModuleCdc.UnmarshalJSON(acknowledgement, ack); err != nil {
+		return sdkerrors.Wrapf(err, "cannot unmarshal ICS-29 incentivized packet acknowledgement: %v", ack)
```

Comment:
I'm not sure of the exact semantics of `UnmarshalJSON`, but given the prevalent practices in Go, would this unmarshalling succeed for _any_ JSON result regardless of whether it contains the expected fields? If so, the error condition would not be triggered unless it contains non-JSON content.

I have been thinking that it may be more practical if the fee middleware can be enabled/disabled for each chain individually without requiring negotiation. Looking at all the protocol interaction, the only requirement for both sides to enable fee simultaneously is so the ack bytes containing the forward relayer addresss can be decoded correctly.

We should instead re-design this such that in case the decoding of the acknowledgement bytes fail, it would simply refund the escrowed receive fee and forward the raw ack to the inner module. Granted, this would also require `UnmarshalJSON` to indeed fail if the JSON raw content do not contain the forward relayer address and result fields.